compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -981,27 +981,50 @@ fn overloaded_place<'a, 'tcx>(
     ExprKind::Deref { arg: ref_expr.to_ref() }
 }
 
-fn capture_upvar<'tcx>(
+fn capture_upvar<'a, 'tcx>(
     cx: &mut Cx<'_, 'tcx>,
     closure_expr: &'tcx hir::Expr<'tcx>,
-    var_hir_id: hir::HirId,
+    captured_place: &'a ty::CapturedPlace<'tcx>,
     upvar_ty: Ty<'tcx>,
 ) -> ExprRef<'tcx> {
-    let upvar_id = ty::UpvarId {
-        var_path: ty::UpvarPath { hir_id: var_hir_id },
-        closure_expr_id: cx.tcx.hir().local_def_id(closure_expr.hir_id),
-    };
-    let upvar_capture = cx.typeck_results().upvar_capture(upvar_id);
+    let upvar_capture = captured_place.info.capture_kind;
     let temp_lifetime = cx.region_scope_tree.temporary_scope(closure_expr.hir_id.local_id);
-    let var_ty = cx.typeck_results().node_type(var_hir_id);
-    let captured_var = Expr {
+    let var_ty = captured_place.place.base_ty;
+
+    let var_hir_id = match captured_place.place.base {
```

Comment:
So...this actually confuses me! I would expect that the base of a captured place might not be an upvar. Isn't this triggered in a case like

```rust
let x = 22;
let c = || drop(x);
```

Here, I'd expect the "place base" for this closure to be `Local(x)`...?

Reply:
I'll add comments here, but this is from within-the closure perspective instead of during closure construction prespective. The place we are matching on was created during Capture Analysis and therefore in that context each of these variables is an upvar.